### PR TITLE
Add recipe for object_store_ffi

### PR DESCRIPTION
**Pull request by nickrobinson251**
- `object_store_ffi` (https://github.com/relationalAI/object_store_ffi) defines a C interface for the Rust crate `object_store` (https://github.com/apache/arrow-rs/tree/master/object_store), which we then use in `ObjectStore.jl` (https://github.com/RelationalAI/ObjectStore.jl/)

### CODE REVIEW

**Comment by nickrobinson251** [comment]
The current error i s

I can't replicate this locally running this script, so i'm wondering if this is something Yggdrasil specific?

For what it is worth, this `build_tarballs.jl` script seemed to be working before moving the Rust code to its own repo and the script to Yggdrasil...
- the script i used before was https://github.com/RelationalAI/ObjectStore.jl/blob/9b0ad2abfde8fbed76cb7258d6623887eab7662c/deps/binary_builder/build_tarballs.jl
- and when run locally that produced the tarballs and JLL package at https://github.com/nickrobinson251/object_store_ffi_jll.jl/

File: O/object_store_ffi/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+version = v"0.1.0"
+
+sources = [
+    # https://github.com/RelationalAI/object_store_ffi/commit/c03a55e65512136c91b6f74787191d34ce2c48b3
+    GitSource("https://github.com/RelationalAI/object_store_ffi.git", "c03a55e65512136c91b6f74787191d34ce2c48b3")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd ${WORKSPACE}/srcdir/
```
Comment (giordano) [inline O/object_store_ffi/build_tarballs.jl:13]: Presumably you want to enter the directory object_store_ffi?

Comment (nickrobinson251) [reply O/object_store_ffi/build_tarballs.jl:13]: :lightbulb: 
```suggestion
cd ${WORKSPACE}/srcdir/object_store_ffi/
```

**Comment by giordano** [comment]
```
[14:52:53] The following warnings were emitted during compilation:
[14:52:53] 
[14:52:53] warning: In file included from /opt/x86_64-linux-musl/registry/src/index.crates.io-6f17d22bba15001f/ring-0.17.7/pregenerated/aesv8-armx-linux64.S:4:0:
[14:52:53] warning: include/ring-core/asm_base.h:73:2: error: #error "ARM assembler must define __ARM_ARCH"
[14:52:53] warning:  #error "ARM assembler must define __ARM_ARCH"
[14:52:53] warning:   ^
[14:52:53] warning: In file included from /opt/x86_64-linux-musl/registry/src/index.crates.io-6f17d22bba15001f/ring-0.17.7/pregenerated/aesv8-armx-linux64.S:7:0:
[14:52:53] warning: include/ring-core/arm_arch.h:82:2: error: #error "ARM assembler must define __ARM_ARCH"
[14:52:53] warning:  #error "ARM assembler must define __ARM_ARCH"
[14:52:53] warning:   ^
[14:52:53] 
[14:52:53] error: failed to run custom build command for `ring v0.17.7`
```
when [targeting aarch64-linux-gnu](https://buildkite.com/julialang/yggdrasil/builds/7636#018ce98e-e6a0-47ff-92a6-b498c775df36/619-1007) is the same as https://github.com/JuliaPackaging/Yggdrasil/pull/7831#issuecomment-1867766030 and it should be fixed by passing `preferred_gcc_version=v"5"` to `build_tarballs()` function.

Instead [the error with x86_64-apple-darwin](https://buildkite.com/julialang/yggdrasil/builds/7636#018ce98e-e69c-4b46-bc84-9d8af562811c/619-1039)
```
[14:54:24]   = note: Undefined symbols for architecture x86_64:
[14:54:24]             "_uv_async_send", referenced from:
[14:54:24]                 __ZN117_$LT$futures_util..stream..futures_unordered..FuturesUnordered$LT$Fut$GT$$u20$as$u20$futures_core..stream..Stream$GT$9poll_next17h1741c2d3e98dc69aE in object_store_ffi.object_store_ffi.dc4c8641f2c99c-cgu.02.rcgu.o
[14:54:24]                 __ZN16object_store_ffi5start28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$28_$u7b$$u7b$closure$u7d$$u7d$17h4da11c03529aeae9E.llvm.8776284286700787704 in object_store_ffi.object_store_ffi.dc4c8641f2c99c-cgu.08.rcgu.o
[14:54:24]           ld: symbol(s) not found for architecture x86_64
[14:54:24]           clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```
is more puzzling.  Where's the reference to a libuv function coming from?

File: O/object_store_ffi/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,40 @@
+using BinaryBuilder
+
+name = "object_store_ffi"
+version = v"0.1.0"
+
+sources = [
+    # https://github.com/RelationalAI/object_store_ffi/commit/b7b55f31d217b0360858ac1fd5471651d3e73784
+    GitSource("https://github.com/RelationalAI/object_store_ffi.git", "b7b55f31d217b0360858ac1fd5471651d3e73784")
```
Comment (nickrobinson251) [inline O/object_store_ffi/build_tarballs.jl:8]: try out https://github.com/RelationalAI/object_store_ffi/pull/1

**Comment by nickrobinson251** [comment]
> should be fixed by passing preferred_gcc_version=v"5" to build_tarballs() function.

thanks -- pushed a commit for that

File: O/object_store_ffi/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    Platform("x86_64",  "macos"),
+]
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libobject_store_ffi", :libobject_store_ffi),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
```
Comment (giordano) [inline O/object_store_ffi/build_tarballs.jl:32]:
```suggestion
dependencies = Dependency[
```

**Comment by nickrobinson251** [comment]
with `preferred_gcc_version=v"5"` the `aarch64-linux-gnu` failure is a little more interesting
https://buildkite.com/julialang/yggdrasil/builds/7638#018cea39-c489-4553-8817-aa6523eeca8c
```
[18:00:41]    Compiling object_store_ffi v0.1.0 (/workspace/srcdir/object_store_ffi)
[18:00:41] error[E0308]: mismatched types
[18:00:41]    --> src/lib.rs:145:30
[18:00:41]     |
[18:00:41] 145 |         self.error_message = c_string.into_raw();
[18:00:41]     |         ------------------   ^^^^^^^^^^^^^^^^^^^ expected `*mut i8`, found `*mut u8`
[18:00:41]     |         |
[18:00:41]     |         expected due to the type of this binding
[18:00:41]     |
[18:00:41]     = note: expected raw pointer `*mut i8`
[18:00:41]                found raw pointer `*mut u8`
[18:00:41] 
[18:00:41] error[E0308]: mismatched types
[18:00:41]    --> src/lib.rs:152:30
[18:00:41]     |
[18:00:41] 152 |         self.error_message = c_string.into_raw();
[18:00:41]     |         ------------------   ^^^^^^^^^^^^^^^^^^^ expected `*mut i8`, found `*mut u8`
[18:00:41]     |         |
[18:00:41]     |         expected due to the type of this binding
[18:00:41]     |
[18:00:41]     = note: expected raw pointer `*mut i8`
[18:00:41]                found raw pointer `*mut u8`
[18:00:41] 
[18:00:41] For more information about this error, try `rustc --explain E0308`.
[18:00:41] error: could not compile `object_store_ffi` (lib) due to 2 previous errors
[18:00:42]  ---> cargo rustc --release --lib --crate-type=cdylib
[18:00:42]  ---> cargo rustc --release --lib --crate-type=cdylib
[18:00:42] Previous command exited with 101
[18:00:42] Child Process exited, exit code 101
ERROR: LoadError: Build for object_store_ffi on aarch64-linux-gnu did not complete successfully
```

But tbh i don't know for sure we need `aarch64-linux` right now, so i'm just going to drop it

**Comment by giordano** [comment]
Well, I found https://doc.rust-lang.org/alloc/ffi/struct.CString.html#method.into_raw which correctly indicates the output of `into_raw` as `c_char`, but now I don't understand the difference between the two documentations, I'm quite confused.

**Comment by nickrobinson251** [comment]
wahoo! that seems to have fixed it! Thanks, Mosè -- i'm new to all this and your help is really appreciated!

Merging this PR will create a JLL package (in JuliaBinaryWrappers), right? there's nothing extra to do to request a JLL package? 

I'm thinking once we merge this PR, i'll `git tag` whatever `object_store_ffi` commit we use as `v0.1.0` -- is that a common workflow? I think ideally we'd tag `object_store_ffi` first, then build from the tag, but i'm thinking at least in the early days of developing the library Yggdrasil might alert us to a few build issues we'll then have to fix before release.

**Comment by giordano** [comment]
> i'm new to all this

I don't have experience with Rust either, I just googled the "Cstring into_raw" :sweat_smile: 

> Merging this PR will create a JLL package (in JuliaBinaryWrappers), right? there's nothing extra to do to request a JLL package?

Yes, no.

> I'm thinking once we merge this PR, i'll git tag whatever object_store_ffi commit we use as v0.1.0 -- is that a common workflow?

Usually we do the other way around, we build on the commit corresponding to an upstream tag :slightly_smiling_face:  But yeah, a tag on the commit we built would be nice to keep things in-sync.

**Pull request merged**